EIPS/eip-6865.md

```diff
@@ -0,0 +1,151 @@
+
+## Motivation
+
+The rapid expansion of the web3.0 ecosystem has unlocked numerous opportunities and innovations. However, this growth has also heightened users' vulnerability to security threats, such as phishing scams. Ensuring that users have a comprehensive understanding of the transactions they sign is crucial for mitigating these risks.
+
+In an attempt to address this issue, we developed an in-house, open-source off-chain SDK for wallets to visualize various protocols. However, we encountered several challenges along the way:
+
+- Scalability: Identifying and understanding all protocols that utilize EIP-712 and their respective business logic is a daunting task, particularly with limited resources. Crafting an off-chain solution for all these protocols is nearly impossible.
+- Reliability: Grasping each protocol's business logic is difficult and may lead to misunderstandings of the actual implementation. This can result in inaccurate visualizations, which could be more detrimental than providing no visualization at all.
+- Maintainability: Offering support for protocols with an off-chain solution is insufficient in a rapidly evolving ecosystem. Protocols frequently upgrade their implementations by extending features or fixing bugs, further complicating the maintenance process.
+
+To overcome these challenges, we propose a standardized, on-chain solution that can accommodate the diverse and ever-changing web3.0 ecosystem. This approach would enhance scalability, reliability, and maintainability by shifting the responsibility of visualizing EIP-712 payloads from the wallets to the protocols themselves. Consequently, wallets can use a consistent and effective approach to EIP-712 message visualization.
+
+The adoption of a universal solution will not only streamline the efforts and reduce the maintenance burden for wallet providers, but it will also allow for faster and more extensive coverage across the ecosystem. This will ultimately result in users gaining a clearer understanding of the transactions they're signing, leading to increased security and an improved overall user experience within the crypto space.
+
+Currently, most of the wallets display something similar to image below
+
+![](../assets/eip-6865/current-EIP-712-signature-wallet-interface.png)
+
+With visualization we can achieve something similar to image below where more insightful details are revealed to user thanks to the structured data returned from the EIP
+
+![](../assets/eip-6865/vision-EIP-712-signature-wallet-interface.png)
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Contracts implementing this proposal MUST include the `visualizeEIP712Message` function in the  `verifyingContract` implementation
+
```

Comment:
Even though this is a technical comment, I'm putting it in this pull request instead of in the discussions-to thread because without an answer in the proposal, I don't feel this specification is complete enough:

How should a wallet link a signature request to a contract? Like if I call `eth_signTypedData`, how does the wallet know which contract to use for `visualizeEIP712Message`? The `verifyingContract` in the domain separator? What if the domain separator doesn't include that field?

Reply:
add more details [here](https://github.com/ethereum/EIPs/pull/6865/commits/928ff07cdbc8bcf82a1491fdcfacef6be1471e95)